### PR TITLE
Integrate Stripe Payment Processing and Customer Management

### PR DESCRIPTION
**Pull request by shettydev**
## Description

This PR introduces Stripe integration into our application to enable payment processing and customer management. 
The major changes that are included in this pr are:

1. A new `StripeModule` with a `StripeService` and `StripeController` to handle interactions with the Stripe API.
2. Functionality for creating Stripe customers and payment intents linked to our `User` entity.
3. Robust error handling and logging to ensure reliability and traceability.
4. Added the `stripe` package to manage communication with the Stripe API.
5. Updated the `docker-compose.yml` to include `REDIS_PORT`, `TYPEORM_PORT`, `TYPEORM_USERNAME`, `TYPEORM_PASSWORD` and `TYPEORM_DATABASE` from the envs

### Few other changes include:
- Prettier formatting changes in the `entity/*` folder.
- Updated root `package.json` workspace to handle all backend projects accordingly.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Verified logging for both successful operations and error scenarios.
- Tested error handling for invalid inputs and missing Stripe API keys.
- Confirmed that integration does not impact existing modules or functionality.

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] Verified compatibility with the `User` entity.
- [x] Verified that new dependencies (`stripe`) are correctly installed and imported.

## Additional context

The `stripe` package (version 17.4.0) was added to facilitate secure and scalable payment operations. It supports customer creation, payment intents, and more.

References:
- [Stripe official documentation](https://docs.stripe.com/api?lang=node)
- Also taken reference to a previous project that I have worked on

This integration aligns with our goal to provide seamless payment capabilities within the application. 
Future enhancements may include support for subscription management, refunds, and webhooks to handle real-time events from Stripe, related to the rider and customer.

### CODE REVIEW

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+   * @param description A human-readable description of the payment intent.
+   * @returns The created Stripe Payment Intent.
+   */
+  @Post("payment-intent")
+  async createPaymentIntent(
+    @Body("userId") userId: number,
+    @Body("amount") amount: number,
+    @Body("currency") currency: string,
+    @Body("description") description: string
+  ): Promise<{ success: boolean; data: Stripe.PaymentIntent }> {
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:41]:
```suggestion
  ): Promise<IApiResponse<Stripe.PaymentIntent>> {
```

Lets keep things consistent across the project.

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  /**
+   * Creates a Stripe Customer for a given user.
+   * @param userId The ID of the user to create the customer for.
+   * @returns The created Stripe Customer.
+   */
+  @Post("customer")
+  async createCustomer(
+    @Body("userId") userId: number
+  ): Promise<{ success: boolean; data: Stripe.Customer }> {
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:19]:
```suggestion
  ):  Promise<IApiResponse<Stripe.Customer>> {
```

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    logger.debug(
+      `StripeController.createCustomer: Creating customer for user ID ${userId}`
+    );
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:22]:
```suggestion
```

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  @Post("payment-intent")
+  async createPaymentIntent(
+    @Body("userId") userId: number,
+    @Body("amount") amount: number,
+    @Body("currency") currency: string,
+    @Body("description") description: string
+  ): Promise<{ success: boolean; data: Stripe.PaymentIntent }> {
+    logger.debug(
+      `StripeController.createPaymentIntent: Creating payment intent for user ID ${userId} with amount ${amount} ${currency}`
+    );
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:44]:
```suggestion
```

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+   * valiates whether user existsm or not, and then creates a Stripe customer.
+   * @param userId The ID of the user to create a customer for.
+   * @returns The created Stripe customer.
+   */
+  async createCustomer(userId: number): Promise<Stripe.Customer> {
+    logger.debug(
+      `StripeService.createCustomer: Creating customer for user ID ${userId}`
+    );
+
+    const user = await dbRepo(User).findOneOrFail({ where: { id: userId } });
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:35]:
```suggestion
    const user = await dbReadRepo(User).findOneOrFail({ where: { id: userId } });
```

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      });
+      logger.info(
+        `StripeService.createCustomer: Customer created with ID ${customer.id}`
+      );
+      return customer;
+    } catch (error) {
+      logger.error(
+        `StripeService.createCustomer: Failed to create customer. Error: ${error.message}`
+      );
+      throw new InternalServerErrorException("Failed to create customer");
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:54]: Define a custom error and classify it as either UserFacingError or not.

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+          userId: user.id.toString(),
+        },
+      });
+      logger.info(
+        `StripeService.createCustomer: Customer created with ID ${customer.id}`
+      );
+      return customer;
+    } catch (error) {
+      logger.error(
+        `StripeService.createCustomer: Failed to create customer. Error: ${error.message}`
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:52]:
```suggestion
        `StripeService.createCustomer: Failed to create customer. Error: ${error}`
```

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    userId: number,
+    amount: number,
+    currency: string = "usd",
+    description: string
+  ): Promise<Stripe.PaymentIntent> {
+    logger.debug(
+      `StripeService.createPaymentIntent: Creating payment intent for user ID ${userId}`
+    );
+
+    const user = await dbRepo(User).findOneOrFail({ where: { id: userId } });
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:77]:
```suggestion
    const user = await dbReadRepo(User).findOneOrFail({ where: { id: userId } });
```

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      });
+      logger.info(
+        `StripeService.createPaymentIntent: Payment intent created with ID ${paymentIntent.id}`
+      );
+      return paymentIntent;
+    } catch (error) {
+      logger.error(
+        `StripeService.createPaymentIntent: Failed to create payment intent. Error: ${error.message}`
+      );
+      throw new InternalServerErrorException("Failed to create payment intent");
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:96]: create a custom error and classify it as UserFacingError or not

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+          userId: user.id.toString(),
+        },
+      });
+      logger.info(
+        `StripeService.createPaymentIntent: Payment intent created with ID ${paymentIntent.id}`
+      );
+      return paymentIntent;
+    } catch (error) {
+      logger.error(
+        `StripeService.createPaymentIntent: Failed to create payment intent. Error: ${error.message}`
```
Comment (theshdb) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:94]:
```suggestion
        `StripeService.createPaymentIntent: Failed to create payment intent. Error: ${error}`
```

Printing the whole stack would be helpful in debugging as it provides brief information.

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts (hunk 2 above)
Comment (abdul-ahad12) [reply apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:19]: For userId we are using AuthGuard.

Here is the code for AuthGuard-
https://github.com/novareach-solutions/movemate/blob/main/apps/backend/backend-main/src/shared/guards/auth.guard.ts

Comment (shettydev) [reply apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:19]: Gotcha

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts (hunk 1 above)
Comment (shettydev) [reply apps/backend/backend-main/src/modules/stripe/stripe.controller.ts]: Done

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts (hunk 3 above)
Comment (shettydev) [reply apps/backend/backend-main/src/modules/stripe/stripe.controller.ts]: Done

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts (hunk 4 above)
Comment (shettydev) [reply apps/backend/backend-main/src/modules/stripe/stripe.controller.ts]: Done

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  constructor(private readonly stripeService: StripeService) {}
+
+  /**
+   * Creates a Stripe Customer for a given user.
+   * @param userId The ID of the user to create the customer for.
+   * @returns The created Stripe Customer.
+   */
+  @Post("customer")
+  async createCustomer(
+    @Body("userId") userId: number,
```
Comment (abdul-ahad12) [inline apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:18]: We get userId from AuthGuard, check PR `https://github.com/novareach-solutions/movemate/pull/74` to have a better understanding

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+   * Creates a Stripe Payment Intent for a given user.
+   * @param userId The ID of the user to create the payment intent for.
+   * @param amount The amount to charge, in the currency's smallest unit (e.g. 100 cents to charge $1.00).
+   * @param currency The 3-character currency code (ISO 4217) for the payment intent.
+   * @param description A human-readable description of the payment intent.
+   * @returns The created Stripe Payment Intent.
+   */
+  @Post("payment-intent")
+  async createPaymentIntent(
+    @Body("userId") userId: number,
```
Comment (abdul-ahad12) [inline apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:38]: Use AuthGuard to get userId

File: apps/backend/backend-main/src/modules/stripe/stripe.service.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        phone: user.phoneNumber,
+        metadata: {
+          userId: user.id.toString(),
+        },
+      });
+      logger.info(
+        `StripeService.createCustomer: Customer created with ID ${customer.id}`,
+      );
+      return customer;
+    } catch (error) {
```
Comment (abdul-ahad12) [inline apps/backend/backend-main/src/modules/stripe/stripe.service.ts:57]: No need of try, catch block here, we have a global exception handler

**Review by abdul-ahad12 — APPROVED** [review]

File: apps/backend/backend-main/src/modules/stripe/stripe.controller.ts (hunk 12 above)
Comment (shettydev) [reply apps/backend/backend-main/src/modules/stripe/stripe.controller.ts:40]: Done. I have added `@UseGuards(AuthGuard)` at the controller level, which will protect all routes in this controller

**Pull request merged**